interfaces/symlinks/backend.go

```diff
@@ -0,0 +1,175 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// symlinks is a backend that ensures that configuration files required by
+// interfaces are present in the system. Currently it works only on classic and
+// modifies the classic rootfs.
+package symlinks
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/timings"
+)
+
+// Backend is responsible for maintaining symlinks cache.
+type Backend struct{}
+
+var _ = interfaces.SecurityBackend(&Backend{})
+
+// Initialize does nothing for this backend.
+func (b *Backend) Initialize(opts *interfaces.SecurityBackendOptions) error {
+	return nil
+}
+
+// Name returns the name of the backend.
+func (b *Backend) Name() interfaces.SecuritySystem {
+	return "symlinks"
+}
+
+// Setup will make the symlinks backend generate the specified symlinks.
+func (b *Backend) Setup(appSet *interfaces.SnapAppSet, opts interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) error {
+	symlinkDirs := map[string]bool{}
+	for _, iface := range repo.AllInterfaces() {
+		if symlnIface, ok := iface.(interfaces.SymlinksUser); ok {
+			for _, d := range symlnIface.Directories() {
+				symlinkDirs[d] = true
+			}
+		}
+	}
+	snapName := appSet.InstanceName()
+	// Get the spec that applies to this snap
+	spec, err := repo.SnapSpecification(b.Name(), appSet, opts)
+	if err != nil {
+		return fmt.Errorf("cannot obtain symlinks specification for snap %q: %s",
+			snapName, err)
+	}
+
+	return b.ensureSymlinks(spec.(*Specification), symlinkDirs)
+}
+
+// Remove removes modules symlinks files specific to a given snap.
+// This method should be called after removing a snap.
+func (b *Backend) Remove(snapName string) error {
+	// If called for the system (snapd) snap, that is possible only in a
+	// classic scenario when all other snaps in the system must have been
+	// removed already to allow the removal of the snapd snap. In that
+	// case, the config files will have already been removed by a Setup
+	// call, so we do not need to do anything here.
+
+	// TODO but this needs to be revisited for when we start supporting
```

Comment:
I think we need to record the created symlinks and be able to remove them.

This also affects the maintenance script.

Reply:
Ideally we want that, and I am playing with the idea of using file trusted extended attributes for this. But I worry about filesystem support in corner cases, although using this backend outside hybrid while not happen initially.

Reply:
I had implemented something like this several years ago where I placed all exported content in /var/lib/snapd/exportfs but this had the requirement that host packages had to be adapted to explicitly look for content in that place.

I agree that looking for symlink target is an option but also one that painfully involves - possibly - scanning the entire file system. I'm not sure we should do that.

Personally I think we should only proceed once we have a viable path of cleaning this stuff up at a cost proportional to the number of created symbolic links.

Reply:
not sure if Alfonso did like we did for configfile, but this should be used always with predeclared directories to place the symlinks in, which means we shouldn't end up in the situation that we need to check the entire filesystem